### ISSUE TITLE
dynamic resource allocation: benchmark scheduler plugin

### ISSUE THREAD

**Issue by pohly** [kubernetes/kubernetes#113701]
### What would you like to be added?

Performance optimizations may be possible.

Other ideas for classic DRA:
- randomize node selection with rand.Perm
- remove sorting of node names

For structured parameters (incomplete list):
- cache compiled CEL expressions
- optimize attribute lookup

### Why is this needed?

Better performance.

**Comment by pohly** [comment]
I intend to use this benchmark to simulate a large number of nodes and how well scheduling works when the number of nodes exceeds the maximum number of entries in the PodScheduling arrays.

**Comment by pohly** [comment]
/remove-lifecycle rotten

This is blocked by needing a solution for https://github.com/kubernetes/kubernetes/issues/113702. As long as the scheduler sleeps while scheduling pods with claims (current situation), measuring incremental improvements that reduce CPU usage won't show any significant differences.

**Comment by pohly** [comment]
As noted above, performance of the scheduler plugin depends on apiserver performance, so I am including that in the analysis. Some of the PRs linked to this issue here are for the apiserver.

One idea that I tried out was to replace `[]string` with `sets.String` in the API of `PodSchedulingContext`. The hope was that operations which work with these fields benefit from fast lookup. In practice, such code isn't run often enough to make up for the more complex memory usage and encoding, so performance gets slightly worse. See https://github.com/pohly/kubernetes/commit/cac17b08b75cc63140d0c72c16af2d95781fc7ad for the attempt.

Another idea is to use `[]string` as base type, but extend it with `Sort/Insert/Has` methods and track whether the content is sorted. `Has` then can do a binary search if sorted, otherwise simply iterate. Might not be worth it, though.

**Comment by pohly** [comment]
Running scheduler_perf as a stress test revealed several circumstances where pods where incorrectly moved to the unschedulable queue and only got scheduled when they were flushed out of that queue by the timeout. This should never be necessary.

To make detection of this more likely, the flush period should be chosen so high that it never triggers while running scheduler_perf.

**Comment by pohly** [comment]
The relevant difference is that pods with no unschedulable plugins now go into the backoff queue. To simulate putting a pod into the unschedulable queue, `pInfo.UnschedulablePlugins` must be set. Because of this change of behavior, a concurrent event is no longer needed when the backoff queue is the intended target:

```patch
diff --git a/pkg/scheduler/internal/queue/scheduling_queue_test.go b/pkg/scheduler/internal/queue/scheduling_queue_test.go
index 13a9a89bb02..b17e5a3c3cc 100644
--- a/pkg/scheduler/internal/queue/scheduling_queue_test.go
+++ b/pkg/scheduler/internal/queue/scheduling_queue_test.go
@@ -2288,6 +2288,8 @@ var (
        }
        popAndRequeueAsUnschedulable = func(t *testing.T, logger klog.Logger, queue *PriorityQueue, pInfo *framework.QueuedPodInfo) {
                // To simulate the pod is failed in scheduling in the real world, Pop() the pod from activeQ before AddUnschedulableIfNotPresent() below.
+               // UnschedulablePlugins will get cleared by Pop, so make a copy first.
+               unschedulablePlugins := pInfo.UnschedulablePlugins.Clone()
                if err := queue.activeQ.Add(queue.newQueuedPodInfo(pInfo.Pod)); err != nil {
                        t.Fatalf("Unexpected error during Add: %v", err)
                }
@@ -2298,6 +2300,8 @@ var (
                if p.Pod != pInfo.Pod {
                        t.Fatalf("Expected: %v after Pop, but got: %v", pInfo.Pod.Name, p.Pod.Name)
                }
+               // Simulate plugins that are waiting for some events.
+               p.UnschedulablePlugins = unschedulablePlugins
                if err := queue.AddUnschedulableIfNotPresent(logger, p, 1); err != nil {
                        t.Fatalf("Unexpected error during AddUnschedulableIfNotPresent: %v", err)
                }
@@ -2314,10 +2318,7 @@ var (
                if p.Pod != pInfo.Pod {
                        t.Fatalf("Expected: %v after Pop, but got: %v", pInfo.Pod.Name, p.Pod.Name)
                }
-
-               // A concurrent event forces the pod into the backoff queue instead of the unschedulable queue.
-               queue.MoveAllToActiveOrBackoffQueue(logger, NodeAdd, nil, nil, nil)
-
+               // When there is no known unschedulable plugin, pods always go to the backoff queue.
                if err := queue.AddUnschedulableIfNotPresent(logger, p, 1); err != nil {
                        t.Fatalf("Unexpected error during AddUnschedulableIfNotPresent: %v", err)
                }
```

**Comment by pohly** [comment]
https://github.com/kubernetes/kubernetes/pull/127277 includes all code optimizations which gave measurable performance improvements. Some additional ideas are in https://github.com/pohly/kubernetes/tree/dra-structured-performance-full but those didn't work out.

Another idea is to remove usage of the finalizer. A quick hack does this:
```patch
diff --git a/pkg/controller/resourceclaim/controller.go b/pkg/controller/resourceclaim/controller.go
index 990dbec2a7a..55dfe95b45e 100644
--- a/pkg/controller/resourceclaim/controller.go
+++ b/pkg/controller/resourceclaim/controller.go
@@ -803,7 +803,7 @@ func (ec *Controller) syncClaim(ctx context.Context, namespace, name string) err
 			// This is a sanity check. There shouldn't be any claims without this
 			// finalizer because there's no longer any other way of allocating claims.
 			// Classic DRA was the alternative earlier.
-			if builtinControllerFinalizer >= 0 {
+			if false && builtinControllerFinalizer >= 0 {
 				// Allocated by scheduler with structured parameters. We can "deallocate"
 				// by clearing the allocation.
 				claim.Status.Allocation = nil
@@ -818,7 +818,7 @@ func (ec *Controller) syncClaim(ctx context.Context, namespace, name string) err
 		// Now also remove the finalizer if it is not needed anymore.
 		// Note that the index may have changed as a result of the UpdateStatus call.
 		builtinControllerFinalizer := slices.Index(claim.Finalizers, resourceapi.Finalizer)
-		if builtinControllerFinalizer >= 0 && claim.Status.Allocation == nil {
+		if false && builtinControllerFinalizer >= 0 && claim.Status.Allocation == nil {
 			claim.Finalizers = slices.Delete(claim.Finalizers, builtinControllerFinalizer, builtinControllerFinalizer+1)
 			if _, err := ec.kubeClient.ResourceV1alpha3().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{}); err != nil {
 				return err
diff --git a/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go b/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
index c2670af5acb..9c4e85c1637 100644
--- a/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
+++ b/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
@@ -798,7 +798,7 @@ func (pl *dynamicResources) Reserve(ctx context.Context, cs *framework.CycleStat
 			// The allocation would be enough. The full object is useful for
 			// debugging, testing and the allocator, so let's make it realistic.
 			claim = claim.DeepCopy()
-			if !slices.Contains(claim.Finalizers, resourceapi.Finalizer) {
+			if false && !slices.Contains(claim.Finalizers, resourceapi.Finalizer) {
 				claim.Finalizers = append(claim.Finalizers, resourceapi.Finalizer)
 			}
 			claim.Status.Allocation = allocation
@@ -941,7 +941,7 @@ func (pl *dynamicResources) bindClaim(ctx context.Context, state *stateData, ind
 
 			// The finalizer needs to be added in a normal update.
 			// If we were interrupted in the past, it might already be set and we simply continue.
-			if !slices.Contains(claim.Finalizers, resourceapi.Finalizer) {
+			if false && !slices.Contains(claim.Finalizers, resourceapi.Finalizer) {
 				claim.Finalizers = append(claim.Finalizers, resourceapi.Finalizer)
 				updatedClaim, err := pl.clientset.ResourceV1alpha3().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{})
 				if err != nil {
```

A real solution is more complicated because the apiserver then must prevent accidentally removing an allocated claim. This probably can be done similar to the delete protection of pods.

Not using the finalizer avoids API calls and should increase performance, but benchmarking doesn't confirm that. Quite the opposite, in one case average throughput degrades considerably:
```
                                                                         │          before.out          │                      after.out                      │
                                                                         │ SchedulingThroughput/Average │ SchedulingThroughput/Average  vs base               │
PerfScheduling/SteadyStateClusterResourceClaimTemplate/empty_100nodes-36                     105.6 ± 4%                    102.8 ±  2%   -2.70% (p=0.041 n=6)
PerfScheduling/SteadyStateClusterResourceClaimTemplate/empty_500nodes-36                     117.8 ± 2%                    111.6 ±  5%   -5.26% (p=0.006 n=6)
PerfScheduling/SteadyStateClusterResourceClaimTemplate/half_100nodes-36                      106.8 ± 5%                    104.4 ±  2%   -2.29% (p=0.009 n=6)
PerfScheduling/SteadyStateClusterResourceClaimTemplate/half_500nodes-36                      125.3 ± 3%                    122.8 ±  3%        ~ (p=0.310 n=6)
PerfScheduling/SteadyStateClusterResourceClaimTemplate/full_100nodes-36                     105.55 ± 3%                    16.16 ± 39%  -84.68% (p=0.002 n=6)
PerfScheduling/SteadyStateClusterResourceClaimTemplate/full_500nodes-36                      120.6 ± 2%                    122.8 ±  1%   +1.87% (p=0.013 n=6)
geomean                                                                                      113.3                         81.44        -28.15%
```

This odd behavior needs further investigation.

**Comment by pohly** [comment]
/remove-lifecycle stale

I'm still interested in understanding better why https://github.com/pohly/kubernetes/tree/dra-structured-performance-full didn't improve performance further.

**Comment by SergeyKanzhelev** [comment]
/close

some banchmarking is done. If we need more, let's open another issue
